File: parse/src/main/java/com/parse/Numbers.java
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse;

/**
 * Static utility methods pertaining to {@link Number} instances.
 */
class Numbers {

    /**
     * Add two {@link Number} instances.
     */
    /* package */
    static Number add(Number first, Number second) {
        if (first instanceof Double || second instanceof Double) {
            return first.doubleValue() + second.doubleValue();
        } else if (first instanceof Float || second instanceof Float) {
            return first.floatValue() + second.floatValue();
        } else if (first instanceof Long || second instanceof Long) {
            return first.longValue() + second.longValue();
        } else if (first instanceof Integer || second instanceof Integer) {
            return first.intValue() + second.intValue();
        } else if (first instanceof Short || second instanceof Short) {
            return first.shortValue() + second.shortValue();
        } else if (first instanceof Byte || second instanceof Byte) {
            return first.byteValue() + second.byteValue();
        } else {
            throw new RuntimeException("Unknown number type.");
        }
    }

    /**
     * Subtract two {@link Number} instances.
     */
    /* package */
    static Number subtract(Number first, Number second) {
        if (first instanceof Double || second instanceof Double) {
            return first.doubleValue() - second.doubleValue();
        } else if (first instanceof Float || second instanceof Float) {
            return first.floatValue() - second.floatValue();
        } else if (first instanceof Long || second instanceof Long) {
            return first.longValue() - second.longValue();
        } else if (first instanceof Integer || second instanceof Integer) {
            return first.intValue() - second.intValue();
        } else if (first instanceof Short || second instanceof Short) {
            return first.shortValue() - second.shortValue();
        } else if (first instanceof Byte || second instanceof Byte) {
            return first.byteValue() - second.byteValue();
        } else {
            throw new RuntimeException("Unknown number type.");
        }
    }

    /**
     * Compare two {@link Number} instances.
     */
    /* package */
    static int compare(Number first, Number second) {
        if (first instanceof Double || second instanceof Double) {
            return (int) Math.signum(first.doubleValue() - second.doubleValue());
        } else if (first instanceof Float || second instanceof Float) {
            return (int) Math.signum(first.floatValue() - second.floatValue());
        } else if (first instanceof Long || second instanceof Long) {
            long diff = first.longValue() - second.longValue();
            return (diff < 0) ? -1 : ((diff > 0) ? 1 : 0);
        } else if (first instanceof Integer || second instanceof Integer) {
            return first.intValue() - second.intValue();
        } else if (first instanceof Short || second instanceof Short) {
            return first.shortValue() - second.shortValue();
        } else if (first instanceof Byte || second instanceof Byte) {
            return first.byteValue() - second.byteValue();
        } else {
            throw new RuntimeException("Unknown number type.");
        }
    }
}

File: parse/src/main/java/com/parse/Numbers.kt
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse

import kotlin.math.sign

/**
 * Static utility methods pertaining to [Number] instances.
 */
internal object Numbers {
    /**
     * Add two [Number] instances.
     */
    /* package */
    @JvmStatic
    fun add(first: Number, second: Number): Number {
        return if (first is Double || second is Double) {
            first.toDouble() + second.toDouble()
        } else if (first is Float || second is Float) {
            first.toFloat() + second.toFloat()
        } else if (first is Long || second is Long) {
            first.toLong() + second.toLong()
        } else if (first is Int || second is Int) {
            first.toInt() + second.toInt()
        } else if (first is Short || second is Short) {
            first.toShort() + second.toShort()
        } else if (first is Byte || second is Byte) {
            first.toByte() + second.toByte()
        } else {
            throw RuntimeException("Unknown number type.")
        }
    }

    /**
     * Subtract two [Number] instances.
     */
    /* package */
    fun subtract(first: Number, second: Number): Number {
        return if (first is Double || second is Double) {
            first.toDouble() - second.toDouble()
        } else if (first is Float || second is Float) {
            first.toFloat() - second.toFloat()
        } else if (first is Long || second is Long) {
            first.toLong() - second.toLong()
        } else if (first is Int || second is Int) {
            first.toInt() - second.toInt()
        } else if (first is Short || second is Short) {
            first.toShort() - second.toShort()
        } else if (first is Byte || second is Byte) {
            first.toByte() - second.toByte()
        } else {
            throw RuntimeException("Unknown number type.")
        }
    }

    /**
     * Compare two [Number] instances.
     */
    /* package */
    @JvmStatic
    fun compare(first: Number, second: Number): Int {
        return if (first is Double || second is Double) {
            sign(first.toDouble() - second.toDouble()).toInt()
        } else if (first is Float || second is Float) {
            sign(first.toFloat() - second.toFloat()).toInt()
        } else if (first is Long || second is Long) {
            val diff = first.toLong() - second.toLong()
            if (diff < 0) -1 else if (diff > 0) 1 else 0
        } else if (first is Int || second is Int) {
            first.toInt() - second.toInt()
        } else if (first is Short || second is Short) {
            first.toShort() - second.toShort()
        } else if (first is Byte || second is Byte) {
            first.toByte() - second.toByte()
        } else {
            throw RuntimeException("Unknown number type.")
        }
    }
}